File: js/function_examples.js
const assert = require('assert');

// Function values can use `=>` or `function`
const square = x => x * x;
const odd = x => Math.abs(x % 2) === 1;
const lessThanTen = function (x) {return x < 10};
const twice = (f, x) => f(f(x));

// An anonymous function call
assert((x => x + 5)(10) === 15);

// We can pass function values to other functions
assert(twice(square, -3) === 81);
assert(twice(x => x + 1, 5) === 7);

// We can create and return new functions on the fly
function compose(f, g) {
  return x => f(g(x));
}
const isOddWhenSquared = compose(odd, square);
assert(isOddWhenSquared(7));
assert(!isOddWhenSquared(0));

// Array functions often take the place of loops
const a = [9, 7, 4, -1, 8];
assert(!a.every(odd));
assert(a.some(odd));
assert(a.every(lessThanTen));
assert.deepStrictEqual(a.filter(odd), [9, 7, -1]);
assert.deepStrictEqual(a.map(square), [81, 49, 16, 1, 64]);

File: dart/function_examples.dart
import 'dart:math' as math;
void main() {
  // We don't have to name functions to call them
  var y = ((x) { return x + 5; })(10);
  print(y);                       // 15

  // We can assign function values to variables
  var square = (x) { return x * x; };
  // Single line functions that return something can be expressed by a shorter
  // syntax. 
  var odd = (x) => (x % 2).abs() == 1;
  var lessThanTen = (x) => x < 10;
  var twice = (f, x) => f(f(x));

  // We can pass function values by name, or anonymously
  print(twice(square, -3));                     // 81
  print(twice((x) => x + 1, 5)); // 7

  // We can create and return new functions on the fly
  compose(f, g) {
    return (x) {
      return f(g(x));
    };
  }
  var isOddWhenSquared = compose(odd, square);
  print(isOddWhenSquared(7));     // true
  print(isOddWhenSquared(0));     // false

  // Several built-in list functions accept functions
  var a = [9, 7, 4, -1, 8];
  print(a.every(odd));            // false
  print(a.any(odd));              // true
  print(a.every(lessThanTen));    // true
  print(a.where(odd));            // [9, 7, -1]
  print(a.map(square));           // [81, 49, 16, 1, 64]

  // Multiple methods can be cascaded on a single object
  a
    ..remove(3)
    ..[2] = 5
    ..retainWhere(odd)
    ..sort();
  print(a);                      // [-1, 5, 7, 9]                     
  
}
